Add `pygeoprocessing.create_raster_from_bounding_box`

phargogh opened the pull request:
This PR adds a new function, `pygeoprocessing.create_raster_from_bounding_box` and refactors `pygeoprocessing.create_raster_from_vector_extents` to use the new function.

Since `create_raster_from_vector_extents` already has extensive functional tests, I just added a test to ensure that the new entrypoint works as expected.

Fixes #276

phargogh committed:
Copying in the implementation from InVEST HRA.

Using https://github.com/natcap/invest/blob/353ff5d3e26948f0d3cc3cd491427a6a9e099fd1/src/natcap/invest/hra.py#L1377

RE:#276

phargogh committed:
Adding type gdal dtype checking.

Correlates with checking in create_raster_from_vector_extents. RE#276

phargogh committed:
Exposing the new function at the package level.

RE:#276

phargogh committed:
Starting a test for the new function.

RE:#276

phargogh committed:
Using new function for create_raster_from_vector_extents.

RE:#276

phargogh committed:
Noting change in HISTORY. RE:#276

emlys commented on src/pygeoprocessing/geoprocessing.py:
```diff
@@ -1034,47 +1034,97 @@ def create_raster_from_vector_extents(
+        y_source = bbox_miny
     raster_transform = [
         x_source, target_pixel_size[0], 0.0,
         y_source, 0.0, target_pixel_size[1]]
```
These trailing 0s could be removed

phargogh replied:
Yep!  Patched in 2dd9cbd

emlys commented on src/pygeoprocessing/geoprocessing.py:
```diff
@@ -1034,47 +1034,97 @@ def create_raster_from_vector_extents(
+        y_source = bbox_maxy
     else:
-        y_source = shp_extent[2]
+        y_source = bbox_miny
```
These lines look like good candidates for the ternary operator, if you'd like to condense them

phargogh replied:
Great idea!  Patched in 0ee2720

emlys commented on src/pygeoprocessing/geoprocessing.py:
```diff
@@ -1034,47 +1034,97 @@ def create_raster_from_vector_extents(
+
+    # Set the nodata value.
+    if target_nodata is not None:
+        band.SetNoDataValue(float(target_nodata))
```
I'm curious why casting to float here?

phargogh replied:
Yeah, this is just for safety, in case the user provides a nodata value with an interesting type, like any of the numpy dtypes, `decimal.Decimal`, etc.  If anything other than an `int` or `float` is provided, there'll be a `TypeError`.  Casting to a `float` will just ensure that the nodata value will be of a type that GDAL can handle.

emlys commented on src/pygeoprocessing/geoprocessing.py:
```diff
@@ -1034,47 +1034,97 @@ def create_raster_from_vector_extents(
 
     # Set the transform based on the upper left corner and given pixel
-    # dimensions
+    # dimensions.  Bounding box is in format [minx, miny, maxx, maxy]
```
Since you've expanded `target_bounding_box` into the four `bbox_*` variables, is this comment still relevant?

phargogh replied:
Nope!  Removed in aed3164

emlys reviewed (approved):
Thanks @phargogh! Just some minor code style suggestions.

phargogh committed:
Being careful to close out the vector. RE:#276

phargogh committed:
Removing unnecessary decimal and trailing 0. RE:#276

phargogh committed:
Condensing conditionals with ternary operators. RE:#276

phargogh committed:
Removing a comment that isn't relevant any more. RE:#276

phargogh commented:
Thanks @emlys !  I've addressed your comments and pushed a couple changes, along with what I hope will fix the test failures on Windows related to gdal objects still being open at test completion.
